pvlib/location.py

```diff
@@ -72,11 +72,12 @@ def __init__(self, latitude, longitude, tz='UTC', altitude=None,
             self.tz = 'UTC'
             self.pytz = pytz.UTC
         elif isinstance(tz, datetime.tzinfo):
+            # This includes pytz timezones.
             self.tz = tz.zone
-            self.pytz = tz
+            self.pytz = pytz.timezone(tz.zone)
         elif isinstance(tz, (int, float)):
-            self.tz = tz
-            self.pytz = pytz.FixedOffset(tz*60)
+            self.tz = f"Etc/GMT{int(-tz):+d}"
```

Comment:
I'm inclined to ask to revert this change. The docstring says `tz` can be a float. There are places that have a non-integer offset from UTC, e.g., the Marquesas are [UTC−09:30](https://en.wikipedia.org/wiki/UTC%E2%88%9209:30)[UTC−09:30](https://en.wikipedia.org/wiki/UTC%E2%88%9209:30). `localize_to_utc` is changed in this PR to examine `Location.pytz` rather than `Location.tz`, so the error we discussed won't occur.

Reply:
There is a point in the Location docstring that says this:

>     Location objects have two timezone attributes:

        * ``tz`` is a IANA timezone string.
        * ``pytz`` is a pytz timezone object.

While I appreciate my changes may break some things, I think this existing docstring suggests a much more sane approach. (Indeed, in the sequel using `zoneinfo`, I propose using one, and only one, internal representation of the timezone, instead of two, potentially with helper functions s to update this internal representation using int/float offsets, etc..)

Reply:
I'm not saying revert from storing a string, but revert the `int` portion. If `tz` is truncated to an integer, how are timezones such as UTC+05:30 represented?

I'm open to doing away with accepting `int` or `float` for `tz`.

Reply:
@cwhanse Sorry I misunderstood your meaning. I was aware that some locations are on half-hour offsets and such, but what is unclear to me is if/how that is represented in `pytz` (or `zoneinfo` for that matter). I do not see any fractional offsets in `zoneinfo.available_timezones()` and this also does not work for me: `pytz.timezone("Etc/GMT+5.5")`.

I must be missing something here. Does something non-standard have to be done to support fractional offsets?

Reply:
> I'm open to doing away with accepting `int` or `float` for `tz`.

Hi again @cwhanse. It looks like the standard "non-integral-hour" offsets are only supported in the IANA standard through place names such as "Asia/Kathmandu" and "Asia/Yangon", but _not_ through `Etc/GMT+X.Y` style offsets.

Given that, I think the most trouble-free thing to do is to remove the `int` and `float` options and be sure to discuss this issue in the documentation tutorial. (Also, the only supported `tz` strings would be the IANA ones.) Sounds like you are amenable to this.

Keeping both fields `tz` (only as a string) and `pytz` (as a `pytz` timezone type), then I think I should add setters that change one when the other is changed after the Location object has already been already initialized.

Reply:
A proposal for the minimum impact on user code: do away with accepting `float` for the parameter `tz`, and keep your translation to a string offset. `Location.tz` is meant to be a string, so the mistake was setting `self.tz = tz` when `tz` is an integer or float.

For a future major version, we can discuss the merits of having only one of `tz` or `pytz`, of keeping both, and of transitioning away from `pytz`. I don't think we want to take steps in that direction in this minor PR that adds tests.